sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,253 @@
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsDefensiveConcurrency(), Configs.getOpenConnectionsDefensiveConcurrency()));
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getOpenConnectionsAggressiveConcurrency(), Configs.getOpenConnectionsAggressiveConcurrency()));
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider, Duration aggressiveConnectionEstablishmentDuration) {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer();
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+        this.aggressivenessHint = new AtomicReference<>(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE);
+        this.endpointsUnderMonitorMap = new ConcurrentHashMap<>();
+        this.openConnectionsHandler = new RntbdOpenConnectionsHandler(endpointProvider);
+        this.endpointProvider = endpointProvider;
+
+        // start as a background task
+        openConnectionBackgroundTask = this.getOpenConnectionsPublisher();
+        Flux<Integer> backgroundOpenConnectionsSinkReInstantiationTask = Flux
+                .just(1)
+                .delayElements(aggressiveConnectionEstablishmentDuration);
+
+        if (aggressiveConnectionEstablishmentDuration != null && aggressiveConnectionEstablishmentDuration.compareTo(Duration.ZERO) > 0) {
+            backgroundOpenConnectionsSinkReInstantiationTask
+                    .doOnComplete(() -> reInstantiateOpenConnectionsPublisherAndSubscribe())
+                    .subscribeOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                    .subscribe();
+        }
+    }
+
+    public ProactiveOpenConnectionsProcessor(final RntbdEndpoint.Provider endpointProvider) {
+        this(endpointProvider, null);
+    }
+
+    public synchronized OpenConnectionTask submitOpenConnectionTaskOutsideLoop(
+        String collectionRid,
+        URI serviceEndpoint,
+        Uri addressUri,
+        int minConnectionsRequiredForEndpoint) {
+
+        OpenConnectionTask openConnectionTask = new OpenConnectionTask(collectionRid, serviceEndpoint, addressUri, minConnectionsRequiredForEndpoint);
+        this.submitOpenConnectionTaskOutsideLoopInternal(openConnectionTask);
+
+        return openConnectionTask;
+    }
+
+    private synchronized void submitOpenConnectionTaskOutsideLoopInternal(OpenConnectionTask openConnectionTask) {
+        String addressUriAsString = openConnectionTask.getAddressUri().getURIAsString();
+
+        this.endpointsUnderMonitorMap.compute(addressUriAsString, (key, taskList) -> {
+            if (taskList == null) {
+                taskList = new ArrayList<>();
+            }
+
+            taskList.add(openConnectionTask);
+            if (taskList.size() == 1) {
```

Comment:
this `taskList` can be mutated by multiple threads, do we really need to only check for 1 here? Can't we check for >= 1 or something like that.

Reply:
Discussed offline.